wallet: Refactor OutputGroups to handle fees and spending eligibility on grouping

achow101 opened the pull request:
Even after #17458, we still deal with setting fees of an `OutputGroup` and filtering the `OutputGroup` outside of the struct. We currently make all of the `OutputGroup`s in `SelectCoins` and then copy and modify them within each `SelectCoinsMinConf` scenario. This PR changes this to constructing the `OutputGroup`s within the `SelectCoinsMinConf` so that the scenario can be taken into account during the group construction. Furthermore, setting of fees and filtering for effective value is moved into `OutputGroup::Insert` itself so that we don't add undesirable outputs to an `OutputGroup` rather than deleting them afterwards.

To facilitate fee calculation and effective value filtering during `OutputGroup::Insert`, `OutputGroup` now takes the feerates in its constructor and computes the fees and effective value for each output during `Insert`.

While removing `OutputGroup`s in accordance with the `CoinEligibilityFilter` still requires creating the `OutputGroup`s first, we can do that within the function that makes them - `GroupOutput`s.

achow101 committed:
Remove OutputGroup non-default constructors

achow101 committed:
Move GroupOutputs into SelectCoinsMinConf

achow101 committed:
Move fee setting of OutputGroup to Insert

OutputGroup will handle the fee and effective value computations
inside of Insert. It now needs to take the effective feerate and long
term feerates as arguments to its constructor.

achow101 committed:
Move EligibleForSpending into GroupOutputs

Instead of filtering after the OutputGroups have been made, do it as
they are being made.

achow101 committed:
Move OutputGroup positive only filtering into Insert

achow101 committed:
Explicitly filter out partial groups when we don't want them

Instead of hacking OutputGroup::m_ancestors to discourage the inclusion
of partial groups via the eligibility filter, add a parameter to the
eligibility filter that indicates whether we want to include the group.
Then for those partial groups, don't return them in GroupOutputs if we
indicate they aren't desired.

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2489,8 +2489,8 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
         (m_spend_zero_conf_change && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, max_ancestors/2, max_descendants/2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && !fRejectLongChains && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max()), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used));
+        (m_spend_zero_conf_change && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1, true), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
```
nit: annotate the new bool

Also why is this being set to true? I thought that we get more *relaxed* as we fail to select with "nicer" coin sets? This seems to make it a tighter criteria?

achow101 replied:
No, it's less restrictive. When it is false, we won't include partial groups. When is true, we do. At least that is the intended behavior.

instagibbs replied:
ahhhh, I had "avoid partial" in my head for this.

instagibbs replied:
so yes, annotation is :ok_hand: because it would have fixed my thought here

achow101 replied:
Annotated

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2489,8 +2489,8 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
-        (m_spend_zero_conf_change && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && !fRejectLongChains && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max()), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used));
+        (m_spend_zero_conf_change && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1, true), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
+        (m_spend_zero_conf_change && !fRejectLongChains && SelectCoinsMinConf(value_to_select, CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max(), std::numeric_limits<uint64_t>::max(), true), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used));
```
nit: annotate the new bool

achow101 replied:
Done

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -4235,9 +4235,9 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
-                // Make this unattractive as we want coin selection to avoid it if possible
-                group.m_ancestors = max_ancestors - 1;
+            if (full_groups.count(it.first) > 0 && !filter.m_include_partial_groups) {
+                // Don't include partial groups if we don't want them yet
```
yet?

achow101 replied:
removed

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -300,16 +300,24 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<OutputGroup>& group
+void OutputGroup::Insert(const CInputCoin& output, int depth, bool from_me, size_t ancestors, size_t descendants, bool positive_only) {
+    // Compute the effective value first
+    CAmount coin_fee = output.m_input_bytes < 0 ? 0 : m_effective_feerate.GetFee(output.m_input_bytes);
+    CAmount ev = output.txout.nValue - coin_fee;
```
Done

achow101 commented on src/wallet/coinselection.cpp:
```diff
@@ -300,16 +300,24 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<OutputGroup>& group
-void OutputGroup::Insert(const CInputCoin& output, int depth, bool from_me, size_t ancestors, size_t descendants) {
+void OutputGroup::Insert(const CInputCoin& output, int depth, bool from_me, size_t ancestors, size_t descendants, bool positive_only) {
+    // Compute the effective value first
+    CAmount coin_fee = output.m_input_bytes < 0 ? 0 : m_effective_feerate.GetFee(output.m_input_bytes);
```
Done

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/20040/commits/f6b305273910db0e46798d361413a7e878cb45f7

fanquake commented:
cc @Xekyo

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -302,6 +302,16 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<OutputGroup>& group
 void OutputGroup::Insert(const CInputCoin& output, int depth, bool from_me, size_t ancestors, size_t descendants) {
     m_outputs.push_back(output);
+    CInputCoin& coin = m_outputs.back();
+    coin.m_fee = coin.m_input_bytes < 0 ? 0 : m_effective_feerate.GetFee(coin.m_input_bytes);
```
I think I have asked this before, but why can `m_input_bytes` ever be below zero here? Perhaps it would be good to have a comment to explain that.

achow101 replied:
`m_input_bytes` is initialized to `-1` to indicate that it hasn't been calculated yet.

Xekyo replied:
But this is at the point where we calculate the effective value of UTXOs, so we need to know the size. Why would we want to mitigate a missing size here rather than throwing?

achow101 replied:
Since the intention is to not change behavior in this PR, I think I will leave this as is for now.

Additionally I don't think it is guaranteed that when we add a `CInputCoin` to an `OutputGroup` that we do know the size. It could be for a preset input or an input not in the wallet (there is a PR for that) where we add those coins to an `OutputGroup` and just don't use the effective value calculation. In those instances, the `m_input_bytes` may not be set.

fjahr replied:
A comment would still be a good idea :)

Xekyo commented on src/wallet/wallet.cpp:
```diff
@@ -4237,8 +4230,10 @@ std::vector<OutputGroup> CWallet::GroupOutputs(const std::vector<COutput>& outpu
+                // This is for if each output gets it's own OutputGroup
+                OutputGroup coin(effective_feerate, long_term_feerate);
+                coin.Insert(input_coin, output.nDepth, output.tx->IsFromMe(ISMINE_ALL), ancestors, descendants);
+                if (coin.EligibleForSpending(filter)) groups.push_back(coin);
```
Wouldn't this mean that if you got a tiny and a large UTXO associated with the same address that you would potentially form a OutputGroup with just the large coin? Shouldn't the group rather be ineligible as a whole to avoid the partial spend?

achow101 replied:
This would make the wallet more vulnerable to dust attacks. An attacker could them lock out a user from their funds by sending dust to an already used address.

Xekyo replied:
Yeah, I guess dust should get ignored altogether, but at higher fee rates this could in the worst-case even affect a small amount and a slightly larger amount, which should be prohibited by the partial spending restriction.

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -300,16 +300,24 @@ bool KnapsackSolver(const CAmount& nTargetValue, std::vector<OutputGroup>& group
+void OutputGroup::Insert(const CInputCoin& output, int depth, bool from_me, size_t ancestors, size_t descendants, bool positive_only) {
+    // Compute the effective value first
+    const CAmount coin_fee = output.m_input_bytes < 0 ? 0 : m_effective_feerate.GetFee(output.m_input_bytes);
+    const CAmount ev = output.txout.nValue - coin_fee;
```
I think that this will result in the OutputGroup accepting uneconomic inputs in the case that the recipient is paying the fees. Should we perhaps rather filter by whether the UTXO are uneconomic, but just calculate with the value instead of the effective value for the case of the recipient paying the fees?

achow101 replied:
Given that the recipient paying the fees is typically used when sweeping the wallet, I don't think it really matters.

Xekyo replied:
Oh, I didn't realize that it was just used for that. Wouldn't it be much easier to implement that as a "send everything": no coin selection, just sum up everything, deduct fees and pay the recipient address that?

achow101 replied:
That's the most common use case I think, but not the only one.

Xekyo replied:
It may be worth considering an explicit `sweepwallet` rpc since I don't think the expectations for "recipient pays fees" and "empty my wallets and send as much as you can" necessarily match. Although, maybe in both cases it would be appropriate not to use uneconomic UTXOs. ;)

Xekyo commented:
Code Review ACK https://github.com/bitcoin/bitcoin/pull/20040/commits/f6b305273910db0e46798d361413a7e878cb45f7

Xekyo commented:
Disclaimer:
@achow101  has walked me throw the PR, I have reviewed it at least twice. I think that the concept makes sense, but I'm not familiar with the wallet code globally and my C++ is somewhat rusty.

achow101 committed:
Rewrite OutputGroups to be clearer and to use scriptPubKeys

Rewrite OutputGroups so that the logic is easier to follow and
understand.

There is a slight behavior change as OutputGroups will be grouped by
scriptPubKey rather than CTxDestination as before. This should have no
effect on users as all addresses are a CTxDestination. However by using
scriptPubKeys, we can correctly group outputs which fall into the
NoDestination case. But we also shouldn't have any NoDestination
outputs.

achow101 commented:
I've added a commit to rewrite `GroupOutputs` based on the comments left in downstream PR review (https://github.com/bitcoin/bitcoin/pull/17331#discussion_r536209337)

Xekyo commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
+
+    // Now we go through the entire map and pull out the OutputGroups
+    for (const auto& groups_pair : groups_map) {
+        const std::vector<OutputGroup>& groups = groups_pair.second;
```
`groups` ⇒ `output_groups_per_spk`

achow101 replied:
Changed to `groups_per_spk`

Xekyo commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
+    // For each COutput, we check if the scriptPubKey is in the map, and if it is, the COutput's CInputCoin is added
+    // to the last OutputGroup in the vector for the scriptPubKey. When the last OutputGroup has
+    // OUTPUT_GROUP_MAX_ENTRIES CInputCoins, a new OutputGroup is added to the end of the vector.
+    std::map<CScript, std::vector<OutputGroup>> groups_map;
```
`groups_map` ⇒ `spk_to_output_groups_map`

achow101 replied:
Changed to `spk_to_groups_map`

Xekyo commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
+    }
+
+    // Now we go through the entire map and pull out the OutputGroups
+    for (const auto& groups_pair : groups_map) {
```
`groups_pair` ⇒ `spk_and_output_groups`

achow101 replied:
Changed to `spk_and_groups_pair`

murchandamus reviewed (commented):
Just reviewed 97256cabb3723dee32dc20aade7d3ae618cff59c

It's much clearer, thank you also for the illustrative comments. Have a few naming suggestions.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
-        if (output.fSpendable) {
-            CTxDestination dst;
-            CInputCoin input_coin = output.GetInputCoin();
+    if (single_coin) {
```
Changed to `separate_coins`

murchandamus reviewed (commented):
Code review ACK: https://github.com/bitcoin/bitcoin/pull/20040/commits/5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
+
+    // Now we go through the entire map and pull out the OutputGroups
+    for (const auto& spk_and_groups_pair: spk_to_groups_map) {
+        const std::vector<OutputGroup>& groups_per_spk= spk_and_groups_pair.second;
```
in 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

whitespace missing before `=`

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
+        const std::vector<OutputGroup>& groups_per_spk= spk_and_groups_pair.second;
+
+        // Go through the vector backwards. This allows for the first item we deal with being the partial group.
+        for (auto group_it = groups_per_spk.rbegin(); group_it != groups_per_spk.rend(); group_it++) {
```
in 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

nit: `++group_it`

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
-            // If the OutputGroup is not eligible, don't add it
+
+            // Check the OutputGroup's eligibility. Only add the eligible ones.
             if (positive_only && group.effective_value <= 0) continue;
```
in 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

same as above, I think this is not necessary since Insert takes care of this now.

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -4193,58 +4193,90 @@ bool CWalletTx::IsImmatureCoinBase() const
+        size_t ancestors, descendants;
+        chain().getTransactionAncestry(output.tx->GetHash(), ancestors, descendants);
+        CInputCoin input_coin = output.GetInputCoin();
+        CScript spk = input_coin.txout.scriptPubKey;
```
in 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

nit: `spk` seems to be only used in the line below so I would drop it.

fjahr commented on src/bench/coin_selection.cpp:
```diff
@@ -42,21 +42,19 @@ static void CoinSelection(benchmark::Bench& bench)
 
-    // Create groups
-    std::vector<OutputGroup> groups;
+    // Create coins
```
in 6148a8acda5e594bb9b3b2d989056f9e03ddbdbd

nit: This comment isn't that helpful, I would suggest something like "Prepare coins in a format that can be passed to SelectCoinsMinConf()"

fjahr commented on src/wallet/wallet.cpp:
```diff
@@ -2381,6 +2381,8 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
         temp.m_confirm_target = 1008;
         CFeeRate long_term_feerate = GetMinimumFeeRate(*this, temp, &feeCalc);
 
+        std::vector<OutputGroup> groups = GroupOutputs(coins, !coin_selection_params.m_avoid_partial_spends, eligibility_filter.max_ancestors);
```
in 6148a8acda5e594bb9b3b2d989056f9e03ddbdbd

This line seems to be the same in both if-else blocks so it could be moved to the beginning of the function before the if

fjahr replied:
Ah, never mind, it changes later when `positive_only` is added

fjahr reviewed (commented):
Code review ACK 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

Nice refactoring! I had a bunch of suggestions for small improvements but I don't consider them blocking for a merge.

One more comment: The behavior change in 416d74fb1687ae1d47a58c153d09d9afe0b6dc60 could be noted a little more explicitly I think. Before this change, the `OutputGroup`s are filled up and then filtered for positive-only later, sometimes then using groups that are not filled completely in coin selection. Now, since the non-positive coins are never inserted we will have full groups with only positive values. This will lead to different results in certain cases. I think this is an improvement and I don't see any issues arise from it but it wouldn't hurt to mention it in the commit message as it is in 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e.

achow101 commented:
@fjahr Since this now has two ACKs, I'm going to leave this as-is for now.

meshcollider reviewed (commented):
Light utACK 5d4597666d589e39354e0d8dd5b2afbe1a5d7d8e

Nice cleanup thanks Andrew!
